File: Questions/187. Repeated DNA Sequences/repeated-dna-sequences.py
"""
Run Envrioment:
Python 3.9

Time and Space Complexity:
Time: O(len(s))
Space: O(len(s))
Actual Time: 68 ms
Actual Space: 27.5 MB
"""

from collections import Counter

def findRepeatedDnaSequences(s: str) -> list[str]:
        if len(s) < 10:
            return []
        
        counter = Counter(s[i: i + 10] for i in range(len(s) - 10 + 1))
        
        return [key for key in counter if counter[key] > 1]


if __name__ == "__main__":
    print(findRepeatedDnaSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"))
    print(findRepeatedDnaSequences("AAAAAAAAAAAAA"))

File: Questions/187. Repeated DNA Sequences/RepeatedDNASequences.java
/* 
Run Envrioment:
Java JDK 14

Time and Space Complexity:
Time: O(len(s))
Space: O(len(s))
Actual Time: 18 ms
Actual Space: 47.6 MB
*/

import java.util.*;

public class RepeatedDNASequences {
    public static List<String> findSequences(String s) {
        if (s.length() < 10) {
            return new ArrayList<String>();
        }

        Set<String> visited = new HashSet<String>();
        Set<String> result = new HashSet<String>();

        for (int i = 0; i < s.length() - 10 + 1; i++) {
            String temp = s.substring(i, i + 10);
            
            if (visited.contains(temp)){
                result.add(temp);
            }

            visited.add(temp);
        }

        return new ArrayList<String>(result);
    }

    public static void main(String[] args){
        System.out.println(RepeatedDNASequences.findSequences("AAAAACCCCCAAAAACCCCCCAAAAAGGGTTT"));
        System.out.println(RepeatedDNASequences.findSequences("AAAAAAAAAAAAA"));
    }
}